pyspecdata/core.py

```diff
@@ -3538,32 +3558,67 @@ def dot(self,arg):
+                dot_dim_B = arg._matmul_along
+            else:
+                dot_dim_B = dot_dim_A
+        elif hasattr(arg,'_matmul_along'):
+            dot_dim_B = arg._matmul_along
+            dot_dim_A = dot_dim_B
+        else:
+            matching_dims = set(self.dimlabels) & set(arg.dimlabels)
+            if len(matching_dims) == 1:
+                dot_dim_A = list(matching_dims)[0]
+                dot_dim_B = dot_dim_A
+            else:
+                raise ValueError("I can't determine the dimension for matrix"
+                        " multiplication!  Either have only one matching dimension,"
+                        " or use .along(")
+        # {{{ unset the "along" setting
+        if hasattr(self,'_matmul_along'):
+            del(self._matmul_along)
+        if hasattr(arg,'_matmul_along'):
+            del(arg._matmul_along)
+        # }}}
+        if dot_dim_B != dot_dim_A:
+            arg = arg.C.rename(dot_dim_B,dot_dim_A)
         A,B = self.aligndata(arg)
-        matching_dims = list(set(self.dimlabels) & set(arg.dimlabels))
-        assert len(matching_dims) > 0, "no matching dimensions!"
         # {{{ store the dictionaries for later use
         axis_coords_dict = A.mkd(A.axis_coords)
         axis_units_dict = A.mkd(A.axis_coords_units)
         axis_coords_error_dict = A.mkd(A.axis_coords_error)
         # }}}
         # manipulate "self" directly
-        self.dimlabels = [j for j in A.dimlabels if j not in matching_dims]
-        match_idx = [A.axn(j) for j in matching_dims]
+        self.dimlabels = [j for j in A.dimlabels if j != dot_dim_A]
         if (self.get_error() is not None) or (arg.get_error() is not None):
             raise ValueError("we plan to include error propagation here, but not yet provided")
-        self.data = tensordot(A.data,B.data,axes=(match_idx,match_idx))
+        ax_idx = A.axn(dot_dim_A) # should be the same for both, since they are aligned
         logger.debug(strm("shape of A is",ndshape(A)))
         logger.debug(strm("shape of B is",ndshape(B)))
-        logger.debug(strm("matching_dims are",matching_dims))
-        newsize = [(A.data.shape[j] if A.data.shape[j] != 1 else B.data.shape[j])
-                for j in range(len(A.data.shape)) if A.dimlabels[j] not in matching_dims]
-        self.data = self.data.reshape(newsize)
+        A.data = A.data[...,newaxis]
+        B.data = B.data[...,newaxis]
+        logger.debug(strm("add a new axis",A.data.shape))
+        logger.debug(strm("add a new axis",B.data.shape))
+        A.data = moveaxis(A.data,ax_idx,-1) # columns position
+        B.data = moveaxis(B.data,ax_idx,-2) # row position
+        logger.debug(strm("after movement",A.data.shape))
+        logger.debug(strm("after movement",B.data.shape))
+        self.data = matmul(A.data,B.data)
+        logger.debug(strm("after mult",self.data.shape))
+        self.data = self.data[...,0,0]
+        logger.debug(strm("remove extras",self.data.shape))
```

Comment:
@bett8500 : so did we establish that matmul and dot were the same speed or different speeds?  What about for 3D and higher vs 2D?